### PR TITLE
Redo `dask.order.order`.  Fix #5584.  Use structural info, not key names

### LINKED ISSUE #5584: Tasks order and scheduling
First of all I apologies because this is more a question than an issue or feature request (though it might turn to one of those in the end).

I am having a really hard time to figure out how to influence task scheduling to get it the way I want it.

Here is my setup. I have two set of tasks, say A and B. Each set has a few thousands tasks in it. A tasks depends on nothing. Each B task depends on a few A tasks output (5 in average).

In my code I delay all A tasks, then I delay B tasks using delayed A objects to model dependencies.

Then I call `client.compute(B_delayed)` and I retrieve results of B on master node on the flow, using `as_completed` (client is from `dask.distributed`).

I would expect the scheduler to figure out that each B task has very few dependencies to A tasks, and therefore to try to complete entire B tasks to be able to clear ressources for A tasks that are no longer needed.

What happens is quite the opposite : the scheduler prioritize A tasks, trying to complete almost all of them while very scarcely computing B tasks (even when a lot of them become ready to run).

This is problematic for us:
- On the flow retrieval is less efficient if all B tasks come at the end (it involves IO operations)
- In some cases, the amount of memory on workers is insufficient to store all A results. This memory could be easily cleared, if only the scheduler decided to complete the corresponding B tasks.

What I tried so far:
- Also submit A tasks with `client.compute()`, with a lower priority (did not change the behavior)
- Submit B tasks in batch with decreasing priority (it helps a bit, but scheduler still wants to complete as many A tasks as possible).
- Mix of both : no noticeable progress

### PR DESCRIPTION
**Pull request by eriknw**
Closes #5584 

This is a rewrite of `dask.order.order`, but the goals remain the same and many previous lessons were taken into consideration.  The new version relies less on the key name by using more metrics and using a different strategy for walking up and down the DAG.

Performance appears to be about the same (often a little faster).

@TomAugspurger also suggested I add some benchmarks to `dask-benchmarks`.

- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`

### CODE REVIEW

**Comment by mrocklin** [comment]
Thanks @eriknw .  When you get back from the holidays, I would be curious about the following:

1.  What were the workloads that motivated this change?
2.  At a high level how does the proposed change change behavior?

The ordering code is fairly core to Dask's functioning today, so changes here should be scrutinized.  It would be good if someone else close to this code (@jcrist @jrbourbeau ?) could review as well.

**Comment by eriknw** [comment]
Certainly.  I expect this to be thoroughly scrutinized and will provide much more supporting material.

1. I began by trying to understand what was happening in `dask.order.order` here: https://github.com/dask/dask/issues/5584#issuecomment-554963958.  The old `order` would alternate up and down the DAG in a sometimes self-avoiding way, which resulted in poor ordering.  This behavior may not show up in other well-patterned DAGs due to good ordering of similar keys.  In the linked issue, the keys were hashes from `dask.delayed`, which revealed this problem that could easily happen elsewhere.

2. At a high level, the proposed change doesn't change any previously expected or established behavior (or at least my understanding of it), but it does the ordering more robustly.  It relies less on using the key and more on using the structure.  The ordering looks good to me in all examples that I looked at from previous issues (but I'm sure I didn't look at or fully understand everything).

Here's a simple example from the tests where the old `order` failed:
```python
def test_avoid_broker_nodes(abcde):
    r"""

    b0    b1  b2
    |      \  /
    a0      a1

    a0 should be run before a1
    """
    a, b, c, d, e = abcde

    ....

    # Switch name of 0, 1 for "b"s too
    dsk = {
        (a, 0): (f,),
        (a, 1): (f,),
        (b, 1): (f, (a, 0)),
        (b, 0): (f, (a, 1)),
        (b, 2): (f, (a, 1)),
    }
    o = order(dsk)
    assert o[(a, 0)] < o[(a, 1)]
```
I found the old approach--although clever and short--to be irredeemable when trying to address these.  I don't mean to sound too critical (the previous way and all the discussions around it were incredibly helpful), so not improving the old way is probably a lack of imagination on my part along with my preference for trying something new.

File: dask/tests/test_order.py [hunk 1]
```diff
@@ -168,11 +175,10 @@ def test_avoid_upwards_branching_complex(abcde):
     }
 
     o = order(dsk)
-
     assert o[(c, 1)] < o[(b, 1)]
 
 
-@pytest.mark.xfail(reason="this case is ambiguous", strict=False)
+# @pytest.mark.xfail(reason="this case is ambiguous", strict=False)  # this now passes
```
Comment (TomAugspurger) [inline dask/tests/test_order.py]: Can probably remove this comment.

Can you update the docstring to reflect the behavior? I see that we run `e` before `d`. Can we say that `b` runs before both `b` and `c`? Or does `b` run first?

File: dask/tests/test_order.py [hunk 2]
```diff
@@ -219,16 +225,15 @@ def test_stacklimit(abcde):
...
 def test_break_ties_by_str(abcde):
     a, b, c, d, e = abcde
     dsk = {("x", i): (inc, i) for i in range(10)}
     x_keys = sorted(dsk)
     dsk["y"] = list(x_keys)
 
     o = order(dsk)
-    expected = {"y": 0}
-    expected.update({k: i + 1 for i, k in enumerate(x_keys)})
+    expected = {"y": 10}
```
Comment (TomAugspurger) [inline dask/tests/test_order.py]: Can you share why this changed?

File: dask/tests/test_order.py [hunk 3]
```diff
@@ -535,3 +541,43 @@ def test_map_overlap(abcde):
     o = order(dsk)
 
     assert o[(b, 1)] < o[(e, 5)] or o[(b, 5)] < o[(e, 1)]
+
+
+def test_use_structure_not_keys(abcde):
+    """ See https://github.com/dask/dask/issues/5584#issuecomment-554963958
```
Comment (TomAugspurger) [inline dask/tests/test_order.py]: Can you add Closes https://github.com/dask/dask/issues/5584 to the original post, or is there remaining work to be done?

File: dask/tests/test_order.py (hunk 2 above)
Comment (eriknw) [reply dask/tests/test_order.py]: It looked to me that the old code here was wrong.  `"y"` cannot be 0.  It has 10 dependencies.

**Comment by eriknw** [comment]
Here is a notebook showing a few examples:

https://gist.github.com/eriknw/0ab7cf29a4f4f6037a22ede14b7f58b8

**Comment by eriknw** [comment]
This is awaiting further review and scrutiny.

As it turns out, I don't think this PR is all that different in spirit and practice than the original.  This makes a few changes how we traverse the graph:
1. We introduce two stacks--an inner and outer--in order to complete the current "tactical target" (an easy-to-compute root) before moving on to the next "tactical target". (There are actually three stacks if you count the pool from which to choose the initial node in of a weakly connected subgraph.)
2. We begin at a leaf node, not a root node.  Although different, this isn't a significant change in and of itself.
3. We use three different key functions--where to start, how to choose a dependency, and how to choose a dependent--and these are designed to try to work together.  I can discuss these keys in more detail upon request.

This PR adds a lot more structure.  Although more complicated, I think this is also more maintainable and modifiable, and I expect to be around a while to address issues that may arise.

In my benchmarks on my machine, this PR is about the same speed or a little faster than the original.  About two thirds of the time is spent in `get_dependencies`, `reverse_dict`, `ndependencies`, and `graph_metrics` (formerly `ndependents`).

**Comment by eriknw** [comment]
Here is the notebook above run using master. There is at least one serious regression!

https://gist.github.com/eriknw/b8478260d6c883cc3c80a1816a993442

Here's the serious regression:

bad_master

We should not be calculation all the dependents on the top at the beginning.  I'd like to add this as a test.

**Comment by eriknw** [comment]
I think the people able to review this are pretty busy this week, so here are relevant previous issues and PRs:
https://github.com/dask/dask/issues/5584
https://github.com/dask/dask/issues/4310
https://github.com/dask/dask/issues/4098
https://github.com/dask/dask/pull/3652
https://github.com/dask/dask/pull/3588
https://github.com/dask/dask-ml/issues/206
https://github.com/dask/dask/issues/3554
https://github.com/dask/dask/issues/3524
https://github.com/dask/dask/pull/3303
https://github.com/dask/dask/pull/3298
https://github.com/dask/dask/pull/3271
https://github.com/dask/dask/pull/3066
https://github.com/dask/dask/pull/3056
https://github.com/dask/dask/issues/3055
https://github.com/dask/dask/pull/3017
https://github.com/dask/dask/issues/3013

Also, here are some benchmarks.

https://gist.github.com/eriknw/82c204b7923f079e267ab245f7228f0a

For these on my machine, the current PR is always faster than master, and sometimes significantly (x3-x10) faster (perhaps when there are many more edges than nodes?).  To be fair, some of the performance improvements of this PR could also be applied to master.

To reiterate, master has had regressions based on previously established and expected behavior.  I believe this PR fixes these.  Okay, I'll try to not revisit this issue until it's been reviewed!

**Comment by TomAugspurger** [comment]
Are you able to reproduce the failing test locally?

```
=================================== FAILURES ===================================

_____________________________ test_describe_empty ______________________________

[gw2] linux -- Python 3.6.7 /home/travis/miniconda/envs/test-environment/bin/python
```

I don't really see how it would be related.

Also, if you're able to, adding benchmarks to https://github.com/dask/dask-benchmarks/ can be done before this is merged. Those should be relatively easy to reveiw.

**Comment by eriknw** [comment]
Aha, I can reproduce the failing test:
```python
import dask
import pandas as pd
import dask.dataframe as dd
df_len0 = pd.DataFrame({"A": [], "B": []})
ddf_len0 = dd.from_pandas(df_len0, 2)
raise_when_computed = ddf_len0.describe(percentiles_method="dask")
```
We expect `raise_when_computed` to, uh, raise when computed.  However, we modify the order the tasks are run, so this the PR raises a different exception.

Here is the DAG on master:
graph-master

and here is the DAG on this PR:
graph-pr

I don't have a strong opinion which DAG to prefer.  I can modify `dependencies_key` to once again include `-total_dependencies[x]` (which is used in master) to go where the work is.  Actually, yeah, let me do that and add this DAG to the order tests.

**Comment by eriknw** [comment]
Okay, so I just tried (and decided against) two things that are in master that aren't part of this PR:
  1. When performing DFS along dependencies, master may decide to proceed to a different dependent.  This is sometimes desirable, but this behavior is also what causes many of the problematic behaviors in master.
    - In this PR, we can do this by creating a new `inner_stack` and appending the old `inner_stack` to a list when we decide there's a better path.  However, I found it was difficult to accurately determine when it's desirable to do this.  Adding this behavior may result in better ordering without significantly impacting the performance, but, for now, I think it's not worth the complexity.
  2. When deciding where to start the next DFS along dependencies, master considers dependents of finished nodes in a LIFO manner.
    - This PR does this in a FIFO manner.  I experimented with deciding between FIFO and LIFO based on some criteria.  This didn't work well.

I just added a new behavior to this PR.  When we encounter a node that has dangling dependents--that is, nodes that are able to be computed and have no dependents--we always compute all of these nodes.  We may then proceed to a different dependent.  This appears to improve the ordering of graphs I've been looking at, and it seems to largely overcome the lack of having the above two behaviors.

I don't think FIFO instead of LIFO behavior will be a significant shortcoming in practice.  In fact, I think FIFO is useful given the lack of (1) above--FIFO gives us a chance to go back and visit branches that we maybe should have gone down instead.  I also think FIFO will work well with the distributed and multithreaded scheduler.

@jcrist I know this is a lot to digest.  Let me know if you want to meet to go over this (and no rush).

File: dask/order.py [hunk 4]
```diff
@@ -108,108 +109,321 @@ def order(dsk, dependencies=None):
...
-    all final dependencies (see module-level comment for more)
+    For each key we return:
+    1.  The number of keys that can only be run after this key is run.  The
+        root nodes have value 1 while deep child nodes will have larger values.
+    2.  The minimum value of the maximum number of dependencies of
+        all final dependencies (see module-level comment for more)
+    3.  The maximum value of the maximum number of dependencies of
+        all final dependencies (see module-level comment for more)
+    4.  The minimum height from a root node
+    5.  The maximum height from a root node
```
Comment (jcrist) [inline dask/order.py]: I find the description of 2 and 3 difficult to understand, perhaps an ascii diagram here would be useful to clarify these metrics?

Comment (eriknw) [reply dask/order.py]: docstring updated.  Hopefully better.

Comment (TomAugspurger) [reply dask/order.py]: @jcrist does this look OK now?

**Comment by jcrist** [comment]
Thanks @eriknw, this looks good to me. Merging!

**Pull request merged**